### PR TITLE
♻️ Migrate amp-web-push example to amp.dev

### LINKED ISSUE #1995: amp-web-push demo failing
## Description of the issue found on ampproject.org

- Page link: https://amp.dev/documentation/examples/components/amp-web-push/
- The demo installs [a ServiceWorker](https://amp.dev/documentation/examples/components/amp-web-push/sw.js), that was developed for AMP by Example version. It requests a local version of the sw-toolbox library, which is not available, causing the registration to fail.

## How do we reproduce the issue?

1. Go to the [demo](https://amp.dev/documentation/examples/components/amp-web-push/) and click on "View demo".
2. Once in the demo page, click on "Subscribe to Notifications".
3. Check the error in console: "Failed to execute 'importScripts' on 'WorkerGlobalScope': The script at 'https://amp.dev/sw-toolbox/sw-toolbox.js' failed to load."

## Proposed fix:

1. Remove the unnecessary code on [sw.js](https://amp.dev/documentation/examples/components/amp-web-push/sw.js) (anything that hasn't to do with Web Push, including sw-toolbox related code).
2. In the amp-web-push component, the new service-worker-scope attribute can be used, to make sure that the SW controls only the path of this demo, without affecting the main SW of amp.dev.

After fixing the SW, the [code of the API](https://glitch.com/~demo-amp-web-push) being called for the web-push service, needs to be updated, to include the proper CORS headers for amp.dev.

### PR DESCRIPTION
**Pull request by lluerich**
Migrated the amp-analytics example backend from [glitch.com]([https://glitch.com/~demo-amp-web-push](https://glitch.com/~demo-amp-web-push) to amp.dev.

It updates the example [here](https://amp.dev/documentation/examples/components/amp-web-push/).

I used the [sw.js](https://glitch.com/edit/#!/demo-amp-web-push?path=public/sw.js) from the example on glitch.com since it no longer uses the deprecated sw-toolbox. Hope that is fine.

Part of the migration/improvements in #2054.

Fixes #1995.

### CODE REVIEW

File: examples/source/1.components/amp-web-push/static/sw.js [hunk 1]
```diff
@@ -256,6 +163,81 @@ function broadcastReply(command, payload) {
...
+self.addEventListener('install', () => self.skipWaiting());
+
+/**
+  Creates the DB to store subscription objects and calls clients.claim(), to allow an active
+  service worker to set itself as the controller for all clients within its scope.
+ */
+self.addEventListener('activate', (event) => {
+  event.waitUntil(
+      idb.open('web-push-db', 1, (upgradeDB) => {
+        upgradeDB.createObjectStore('web-push-subcription', {
```
Comment (sebastianbenz) [inline examples/source/1.components/amp-web-push/static/sw.js]: extract 'web-push-subcription' into a constant

File: examples/source/1.components/amp-web-push/static/sw.js [hunk 2]
```diff
@@ -256,6 +163,81 @@ function broadcastReply(command, payload) {
...
+
+/**
+  Intercepts the /send-push request: retrieves the subscription object from indexedDB, and passes it to an endpoint
+  so WebPush messages can be sent.
+  Any other request goes to the network directly.
+ */
+self.addEventListener('fetch', (event) => {
+  if (event.request.url.includes('send-push')) {
+    event.respondWith(
+        idb.open('web-push-db', 1).then((db) => {
```
Comment (sebastianbenz) [inline examples/source/1.components/amp-web-push/static/sw.js]: it'd be nice to use async/await everywhere

File: examples/source/1.components/amp-web-push/static/sw.js [hunk 3]
```diff
@@ -256,6 +163,81 @@ function broadcastReply(command, payload) {
...
+          const tx = db.transaction(['web-push-subcription'], 'readonly');
+          const store = tx.objectStore('web-push-subcription');
+
+          return store.get(1).then((subscriptionJSON) => {
```
Comment (sebastianbenz) [inline examples/source/1.components/amp-web-push/static/sw.js]: magic number `1`

File: examples/source/1.components/amp-web-push/static/sw.js [hunk 4]
```diff
@@ -256,6 +163,81 @@ function broadcastReply(command, payload) {
...
+  );
+});
+
+/**
+  Listens to push events, and displays a notification, using the payload text.
+ */
+self.addEventListener('push', (event) => {
+  const options = {
+    body: event.data.text(),
+    icon: 'images/notification-flat.png',
```
[the remark this answers is not in the document]
Comment (lluerich) [reply examples/source/1.components/amp-web-push/static/sw.js]: Nope, as of now I just removed it.

File: examples/source/1.components/amp-web-push/static/sw.js [hunk 5]
```diff
@@ -256,6 +163,81 @@ function broadcastReply(command, payload) {
...
+  );
+});
+
+/**
+  Intercepts the /send-push request: retrieves the subscription object from indexedDB, and passes it to an endpoint
+  so WebPush messages can be sent.
+  Any other request goes to the network directly.
+ */
+self.addEventListener('fetch', (event) => {
+  if (event.request.url.includes('send-push')) {
```
[the remark this answers is not in the document]
Comment (lluerich) [reply examples/source/1.components/amp-web-push/static/sw.js]: PTAL

File: examples/source/1.components/amp-web-push/static/sw.js [hunk 6]
```diff
@@ -256,6 +163,81 @@ function broadcastReply(command, payload) {
...
+ */
+self.addEventListener('install', () => self.skipWaiting());
+
+/**
+  Creates the DB to store subscription objects and calls clients.claim(), to allow an active
+  service worker to set itself as the controller for all clients within its scope.
+ */
+self.addEventListener('activate', (event) => {
+  event.waitUntil(
+      idb.open('web-push-db', 1, (upgradeDB) => {
```
[the remark this answers is not in the document]
Comment (lluerich) [reply examples/source/1.components/amp-web-push/static/sw.js]: PTAL

File: examples/source/1.components/amp-web-push/static/sw.js (hunk 1 above)
Comment (lluerich) [reply examples/source/1.components/amp-web-push/static/sw.js]: PTAL

File: examples/source/1.components/amp-web-push/static/sw.js (hunk 3 above)
Comment (lluerich) [reply examples/source/1.components/amp-web-push/static/sw.js]: Moved to constant, PTAL

File: examples/source/1.components/amp-web-push/static/sw.js (hunk 2 above)
Comment (lluerich) [reply examples/source/1.components/amp-web-push/static/sw.js]: PTAL

File: examples/source/1.components/amp-web-push/static/sw.js [hunk 7]
```diff
@@ -256,33 +166,106 @@ function broadcastReply(command, payload) {
...
+      const options = {
+        method: 'POST',
+        headers: {
+          'Accept': 'application/json',
+          'Content-Type': 'application/json',
+        },
+        body: subscriptionJSON.data,
+      };
+
+      return await fetch(self.location.origin +
```
Comment (sebastianbenz) [inline examples/source/1.components/amp-web-push/static/sw.js]: The `await` is not needed here. It's fine to return a promise.

File: examples/source/1.components/amp-web-push/static/sw.js [hunk 8]
```diff
@@ -256,33 +166,106 @@ function broadcastReply(command, payload) {
...
+            '/documentation/examples/components/amp-web-push/send-push', options);
+    })());
+  } else {
+    event.respondWith((async () => {
+      return await fetch(event.request);
```
Comment (sebastianbenz) [inline examples/source/1.components/amp-web-push/static/sw.js]: The await is not needed here. It's fine to return a promise.

File: examples/source/1.components/amp-web-push/static/sw.js (hunk 7 above)
Comment (lluerich) [reply examples/source/1.components/amp-web-push/static/sw.js]: PTAL

File: examples/source/1.components/amp-web-push/static/sw.js (hunk 8 above)
Comment (lluerich) [reply examples/source/1.components/amp-web-push/static/sw.js]: PTAL

**Comment by sebastianbenz** [comment]
Thanks! Can you please merge package.json and run `npm install`

**Comment by lluerich** [comment]
I merged future and resolved the conflict.

**Comment by sebastianbenz** [comment]
thanks!

**Pull request merged**